notification-server/src/processor.rs

```diff
@@ -0,0 +1,103 @@
+use concordium_rust_sdk::{
+    base::contracts_common::AccountAddress,
+    cis2,
+    cis2::Event,
+    types,
+    types::{
+        AccountTransactionEffects, Address, Address::Account,
+        BlockItemSummaryDetails::AccountTransaction, ContractTraceElement,
+    },
+};
+use futures::{Stream, StreamExt};
+use num_bigint::BigInt;
+use std::fmt::Debug;
+
+#[derive(Debug)]
+pub struct NotificationInformation {
+    pub address: AccountAddress,
+    pub amount:  BigInt,
+}
+
+impl NotificationInformation {
+    pub fn new(address: AccountAddress, amount: BigInt) -> Self { Self { address, amount } }
```

Comment:
I disagree. I see no apparent reason why we would need two different values indicating the same thing.

As long as the value is big enough it will be able to be represent either token or CCD.

Reply:
We would need a way to distinguish between `tokensAmount` and `CCDAmount`. If you want to keep `BigInt`, I suggest adding the `ContractAddress` and `TokenID` as options, and that way we can distinguish if your notification is about a token or a CCD amount. 

Small changes in the above code:
- renamed `get_cis2_events_addresses` to `get_cis2_events_addresses`.
- add `contract_address` and `token_id` to the notification.
- removed burn event (if they goal of the notification service is to notify only about increase in balance of an account)
```

#[derive(Debug)]
pub struct NotificationInformation {
    pub address: AccountAddress,
    pub amount: BigInt,
    pub contract_address: Option<ContractAddress>,
    pub token_id: Option<cis2::TokenId>,
}

impl NotificationInformation {
    pub fn new(
        address: AccountAddress,
        amount: BigInt,
        contract_address: Option<ContractAddress>,
        token_id: Option<cis2::TokenId>,
    ) -> Self {
        Self { address, amount,contract_address,token_id }
    }
}

fn convert<T: Into<BigInt>>(
    address: Address,
    amount: T,
    contract_address: Option<ContractAddress>,
    token_id: Option<cis2::TokenId>,
) -> Option<NotificationInformation> {
    let amount: BigInt = amount.into();

    match address {
        Account(address) => Some(NotificationInformation::new(
            address,
            amount,
            contract_address,
            token_id,
        )),
        _ => None,
    }
}

fn get_events(effects: &AccountTransactionEffects) -> Vec<NotificationInformation> {
    match &effects {
...
        AccountTransactionEffects::ContractUpdateIssued { effects } => effects
            .iter()
            .flat_map(|effect| match effect {
                ContractTraceElement::Transferred { to, amount, .. } => {
                    vec![NotificationInformation::new(
                        *to,
                        amount.micro_ccd.into(),
                        None,
                        None,
                    )]
                }
                ContractTraceElement::Updated { data } => data
                    .events
                    .iter()
                    .filter_map(|event| {
                        let contract_address = data.address;

                        match cis2::Event::try_from(event) {
                            Ok(Event::Transfer {
                                to,
                                amount,
                                token_id,
                                ..
                            }) => convert(to, amount.0,  Some(contract_address), Some(token_id)),
                            Ok(Event::Mint {
                                owner,
                                amount,
                                token_id,
                            }) => convert(owner, amount.0,  Some(contract_address), Some(token_id)),
                            _ => None,
                        }
                    })
                    .collect(),
                _ => vec![],
            })
            .collect(),

...
}
```

Reply:
I believe this is a fine initiative, and we definitely do need the token_id as well as the extraction. To lower the impact of this PR I have decided to put that out of scope for this PR.

I have created a task: https://linear.app/concordium/issue/MIS-74/derive-token-id

Reply:
ok, moving the following action points to the next PR:

- rename `get_cis2_events_addresses` function
- add `contract_address` and `token_id` to the notification.
- removed burn event (https://concordium.slack.com/archives/C078CQL92K1/p1719298721127089?thread_ts=1719236927.350389&cid=C078CQL92K1)

Reply:
Using an `enum` seems a more Rust way:

```
#[derive(Debug)]
pub struct CCDAmount {
    pub account: AccountAddress,
    pub amount: Amount,
}

#[derive(Debug)]
pub struct TokenAmount {
    pub account: AccountAddress,
    pub amount: cis2::TokenAmount,
    pub contract_address: ContractAddress,
    pub token_id: cis2::TokenId,
}

fn convert(
    address: Address,
    amount: cis2::TokenAmount,
    contract_address: ContractAddress,
    token_id: cis2::TokenId,
) -> Option<Notification> {
    match address {
        Account(account) => Some(Notification::TokenAmount(TokenAmount {
            account,
            amount,
            contract_address,
            token_id,
        })),
        _ => None,
    }
}

fn get_events(effects: &AccountTransactionEffects) -> Vec<Notification> {
    match &effects {
        AccountTransactionEffects::AccountTransfer { to, amount } => {
            vec![Notification::CCDAmount(CCDAmount {
                account: *to,
                amount: *amount,
            })]
        }
        AccountTransactionEffects::AccountTransferWithMemo { to, amount, .. } => {
            vec![Notification::CCDAmount(CCDAmount {
                account: *to,
                amount: *amount,
            })]
        }
        AccountTransactionEffects::ContractUpdateIssued { effects } => effects
            .iter()
            .flat_map(|effect| match effect {
                ContractTraceElement::Transferred { to, amount, .. } => {
                    vec![Notification::CCDAmount(CCDAmount {
                        account: *to,
                        amount: *amount,
                    })]
                }
                ContractTraceElement::Updated { data } => data
                    .events
                    .iter()
                    .filter_map(|event| {
                        let contract_address = data.address;

                        match cis2::Event::try_from(event) {
                            Ok(Event::Transfer {
                                to,
                                amount,
                                token_id,
                                ..
                            }) => convert(to, amount, contract_address, token_id),
                            Ok(Event::Mint {
                                owner,
                                amount,
                                token_id,
                            }) => convert(owner, amount, contract_address, token_id),
                            Ok(Event::Burn {
                                owner,
                                amount,
                                token_id,
                            }) => convert(owner, amount, contract_address, token_id),
                            _ => None,
                        }
                    })
                    .collect(),
                _ => vec![],
            })
            .collect(),


            // ...


        _ => vec![],
    }
}#[derive(Debug)]
pub struct CCDAmount {
    pub account: AccountAddress,
    pub amount: Amount,
}

#[derive(Debug)]
pub struct TokenAmount {
    pub account: AccountAddress,
    pub amount: cis2::TokenAmount,
    pub contract_address: ContractAddress,
    pub token_id: cis2::TokenId,
}

fn convert(
    address: Address,
    amount: cis2::TokenAmount,
    contract_address: ContractAddress,
    token_id: cis2::TokenId,
) -> Option<Notification> {
    match address {
        Account(account) => Some(Notification::TokenAmount(TokenAmount {
            account,
            amount,
            contract_address,
            token_id,
        })),
        _ => None,
    }
}

fn get_events(effects: &AccountTransactionEffects) -> Vec<Notification> {
    match &effects {
        AccountTransactionEffects::AccountTransfer { to, amount } => {
            vec![Notification::CCDAmount(CCDAmount {
                account: *to,
                amount: *amount,
            })]
        }
        AccountTransactionEffects::AccountTransferWithMemo { to, amount, .. } => {
            vec![Notification::CCDAmount(CCDAmount {
                account: *to,
                amount: *amount,
            })]
        }
        AccountTransactionEffects::ContractUpdateIssued { effects } => effects
            .iter()
            .flat_map(|effect| match effect {
                ContractTraceElement::Transferred { to, amount, .. } => {
                    vec![Notification::CCDAmount(CCDAmount {
                        account: *to,
                        amount: *amount,
                    })]
                }
                ContractTraceElement::Updated { data } => data
                    .events
                    .iter()
                    .filter_map(|event| {
                        let contract_address = data.address;

                        match cis2::Event::try_from(event) {
                            Ok(Event::Transfer {
                                to,
                                amount,
                                token_id,
                                ..
                            }) => convert(to, amount, contract_address, token_id),
                            Ok(Event::Mint {
                                owner,
                                amount,
                                token_id,
                            }) => convert(owner, amount, contract_address, token_id),
                            Ok(Event::Burn {
                                owner,
                                amount,
                                token_id,
                            }) => convert(owner, amount, contract_address, token_id),
                            _ => None,
                        }
                    })
                    .collect(),
                _ => vec![],
            })
            .collect(),


            ...


        _ => vec![],
    }
}
```